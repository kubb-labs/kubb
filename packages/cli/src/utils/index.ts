export * from './getConfig.ts'
export * from './getPlugins.ts'
export * from './watcher.ts'
export * from './getCosmiConfig.ts'
export * from './parseHrtimeToSeconds.ts'
export * from './parseText.ts'
