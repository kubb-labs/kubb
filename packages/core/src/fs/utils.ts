import { normalize, relative } from 'node:path'

function slash(path: string, platform: 'windows' | 'mac' | 'linux' = 'linux') {
  const isWindowsPath = /^\\\\\?\\/.test(path)
  const normalizedPath = normalize(path)

  if (['linux', 'mac'].includes(platform) && !isWindowsPath) {
    // linux and mac
    return normalizedPath.replaceAll(/\\/g, '/').replace('../', '')
  }

  // windows
  return normalizedPath.replaceAll(/\\/g, '/').replace('../', '')
}

export function getRelativePath(rootDir?: string | null, filePath?: string | null, platform: 'windows' | 'mac' | 'linux' = 'linux'): string {
  if (!rootDir || !filePath) {
    throw new Error(`Root and file should be filled in when retrieving the relativePath, ${rootDir || ''} ${filePath || ''}`)
  }

  const relativePath = relative(rootDir, filePath)

  // On Windows, paths are separated with a "\"
  // However, web browsers use "/" no matter the platform
  const slashedPath = slash(relativePath, platform)

  if (slashedPath.startsWith('../')) {
    return slashedPath
  }

  return `./${slashedPath}`
}
