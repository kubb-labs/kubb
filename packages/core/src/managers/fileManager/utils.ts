import path from 'node:path'

import { createExportDeclaration, createImportDeclaration, print } from '@kubb/parser'

import isEqual from 'lodash.isequal'
import { orderBy } from 'natural-orderby'

import { TreeNode } from '../../utils/index.ts'

import type { TreeNodeOptions } from '../../utils/index.ts'
import type { KubbFile } from './types.ts'

type TreeNodeData = { type: KubbFile.Mode; path: KubbFile.Path; name: string }

export type IndexesOptions = {
  treeNode?: TreeNodeOptions
  isTypeOnly?: boolean
  filter?: (file: KubbFile.File) => boolean
  map?: (file: KubbFile.File) => KubbFile.File
  includeExt?: boolean
  output?: string
}

export function getIndexes(
  root: string,
  extName?: KubbFile.Extname,
  { treeNode = {}, isTypeOnly, filter, map, output, includeExt }: IndexesOptions = {},
): Array<KubbFile.File> | null {
  const extMapper: Record<KubbFile.Extname, TreeNodeOptions> = {
    '.ts': {
      extensions: /\.ts/,
      exclude: [/schemas/, /json/],
    },
    '.json': {
      extensions: /\.json/,
      exclude: [],
    },
  }
  const tree = TreeNode.build<TreeNodeData>(root, { ...(extMapper[extName as keyof typeof extMapper] || {}), ...treeNode })

  if (!tree) {
    return null
  }

  const fileReducer = (files: Array<KubbFile.File>, currentTree: typeof tree) => {
    if (!currentTree.children) {
      return []
    }

    if (currentTree.children?.length > 1) {
      const indexPath: KubbFile.Path = path.resolve(currentTree.data.path, 'index.ts')
      const exports: KubbFile.Export[] = currentTree.children
        .filter(Boolean)
        .map((file) => {
          const importPath: string = file.data.type === 'directory' ? `./${file.data.name}` : `./${file.data.name.replace(/\.[^.]*$/, '')}`

          // TODO weird hacky fix
          if (importPath.includes('index') && indexPath.includes('index')) {
            return undefined
          }

          return {
            path: includeExt ? (file.data.type === 'directory' ? `${importPath}/index${extName}` : `${importPath}${extName}`) : importPath,
            isTypeOnly,
          } as KubbFile.Export
        })
        .filter(Boolean)

      files.push({
        path: indexPath,
        baseName: 'index.ts',
        source: '',
        exports: output
          ? exports?.filter((item) => {
            return item.path.endsWith(output.replace(/\.[^.]*$/, ''))
          })
          : exports,
      })
    } else {
      currentTree.children?.forEach((child) => {
        const indexPath = path.resolve(currentTree.data.path, 'index.ts')
        const importPath = child.data.type === 'directory' ? `./${child.data.name}` : `./${child.data.name.replace(/\.[^.]*$/, '')}`

        const exports = [
          {
            path: includeExt ? (child.data.type === 'directory' ? `${importPath}/index${extName}` : `${importPath}${extName}`) : importPath,
            isTypeOnly,
          },
        ]

        files.push({
          path: indexPath,
          baseName: 'index.ts',
          source: '',
          exports: output
            ? exports?.filter((item) => {
              return item.path.endsWith(output.replace(/\.[^.]*$/, ''))
            })
            : exports,
        })
      })
    }

    currentTree.children.forEach((childItem) => {
      fileReducer(files, childItem)
    })

    return files
  }

  const files = fileReducer([], tree).reverse()

  const filteredFiles = filter ? files.filter(filter) : files

  return map ? filteredFiles.map(map) : filteredFiles
}

export function combineFiles(files: Array<KubbFile.File | null>): Array<KubbFile.File> {
  return files.filter(Boolean).reduce((acc, curr: KubbFile.File) => {
    const prevIndex = acc.findIndex((item) => item.path === curr.path)

    if (prevIndex !== -1) {
      const prev = acc[prevIndex]

      if (prev) {
        acc[prevIndex] = {
          ...curr,
          source: prev.source && curr.source ? `${prev.source}\n${curr.source}` : '',
          imports: [...(prev.imports || []), ...(curr.imports || [])],
          exports: [...(prev.exports || []), ...(curr.exports || [])],
          env: { ...(prev.env || {}), ...(curr.env || {}) },
        }
      }
    } else {
      acc.push(curr)
    }

    return acc
  }, [] as Array<KubbFile.File>)
}
/**
 * Support for js, ts and tsx(React)
 */

export const extensions: Array<KubbFile.Extname> = ['.js', '.ts', '.tsx']

export function isExtensionAllowed(baseName: string): boolean {
  return extensions.some((extension) => baseName.endsWith(extension))
}

export function combineExports(exports: Array<KubbFile.Export>): Array<KubbFile.Export> {
  const combinedExports = orderBy(exports, [(v) => !v.isTypeOnly], ['asc']).reduce((prev, curr) => {
    const name = curr.name
    const prevByPath = prev.findLast((imp) => imp.path === curr.path)
    const prevByPathAndIsTypeOnly = prev.findLast((imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly)

    if (prevByPathAndIsTypeOnly) {
      // we already have an export that has the same path but uses `isTypeOnly` (export type ...)
      return prev
    }

    const uniquePrev = prev.findLast(
      (imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly === curr.isTypeOnly && imp.asAlias === curr.asAlias,
    )

    if (uniquePrev || (Array.isArray(name) && !name.length) || (prevByPath?.asAlias && !curr.asAlias)) {
      return prev
    }

    if (!prevByPath) {
      return [
        ...prev,
        {
          ...curr,
          name: Array.isArray(name) ? [...new Set(name)] : name,
        },
      ]
    }

    if (prevByPath && Array.isArray(prevByPath.name) && Array.isArray(curr.name) && prevByPath.isTypeOnly === curr.isTypeOnly) {
      prevByPath.name = [...new Set([...prevByPath.name, ...curr.name])]

      return prev
    }

    return [...prev, curr]
  }, [] as Array<KubbFile.Export>)

  return orderBy(combinedExports, [(v) => !v.isTypeOnly, (v) => v.asAlias], ['desc', 'desc'])
}

export function combineImports(imports: Array<KubbFile.Import>, exports: Array<KubbFile.Export>, source?: string): Array<KubbFile.Import> {
  const combinedImports = orderBy(imports, [(v) => !v.isTypeOnly], ['asc']).reduce((prev, curr) => {
    let name = Array.isArray(curr.name) ? [...new Set(curr.name)] : curr.name

    const hasImportInSource = (importName: string) => {
      if (!source) {
        return true
      }

      const checker = (name?: string) => name && !!source.includes(name)
      return checker(importName) || exports.some(({ name }) => (Array.isArray(name) ? name.some(checker) : checker(name)))
    }

    if (Array.isArray(name)) {
      name = name.filter((item) => hasImportInSource(item))
    }

    const prevByPath = prev.findLast((imp) => imp.path === curr.path && imp.isTypeOnly === curr.isTypeOnly)
    const uniquePrev = prev.findLast((imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly === curr.isTypeOnly)
    const prevByPathNameAndIsTypeOnly = prev.findLast((imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly)

    if (prevByPathNameAndIsTypeOnly) {
      // we already have an export that has the same path but uses `isTypeOnly` (import type ...)
      return prev
    }

    if (uniquePrev || (Array.isArray(name) && !name.length)) {
      return prev
    }

    if (!prevByPath) {
      return [
        ...prev,
        {
          ...curr,
          name,
        },
      ]
    }

    if (prevByPath && Array.isArray(prevByPath.name) && Array.isArray(name) && prevByPath.isTypeOnly === curr.isTypeOnly) {
      prevByPath.name = [...new Set([...prevByPath.name, ...name])]

      return prev
    }

    if (!Array.isArray(name) && name && !hasImportInSource(name)) {
      return prev
    }

    return [...prev, curr]
  }, [] as Array<KubbFile.Import>)

  return orderBy(combinedImports, [(v) => !v.isTypeOnly], ['desc'])
}

export function createFileSource(file: KubbFile.File): string {
  if (!isExtensionAllowed(file.baseName)) {
    return file.source
  }

  const exports = file.exports ? combineExports(file.exports) : []
  const imports = file.imports ? combineImports(file.imports, exports, file.source) : []

  const importNodes = imports.map((item) => createImportDeclaration({ name: item.name, path: item.path, isTypeOnly: item.isTypeOnly }))
  const exportNodes = exports.map((item) => createExportDeclaration({ name: item.name, path: item.path, isTypeOnly: item.isTypeOnly, asAlias: item.asAlias }))

  return [print([...importNodes, ...exportNodes]), getEnvSource(file.source, file.env)].join('\n')
}

type SearchAndReplaceOptions = {
  text: string
  replaceBy: string
  prefix?: string
  key: string
  searchValues?: (prefix: string, key: string) => Array<RegExp | string>
}

function searchAndReplace(options: SearchAndReplaceOptions): string {
  const { text, replaceBy, prefix = '', key } = options

  const searchValues = options.searchValues?.(prefix, key) || [
    `${prefix}["${key}"]`,
    `${prefix}['${key}']`,
    `${prefix}[\`${key}\`]`,
    `${prefix}"${key}"`,
    `${prefix}'${key}'`,
    `${prefix}\`${key}\``,
    new RegExp(`${prefix}${key}`, 'g'),
  ]

  return searchValues.reduce((prev, searchValue) => {
    return prev.toString().replaceAll(searchValue, replaceBy)
  }, text) as string
}

function getEnvSource(source: string, env: NodeJS.ProcessEnv | undefined): string {
  if (!env) {
    return source
  }

  const keys = Object.keys(env)

  if (!keys.length) {
    return source
  }

  return keys.reduce((prev, key: string) => {
    const environmentValue = env[key]
    const replaceBy = environmentValue ? `'${environmentValue.replaceAll('"', '')?.replaceAll("'", '')}'` : 'undefined'

    if (key.toUpperCase() !== key) {
      throw new TypeError(`Environment should be in upperCase for ${key}`)
    }

    if (typeof replaceBy === 'string') {
      prev = searchAndReplace({ text: prev.replaceAll(`process.env.${key}`, replaceBy), replaceBy, prefix: 'process.env', key })
      // removes `declare const ...`
      prev = searchAndReplace({ text: prev.replaceAll(new RegExp(`(declare const).*\n`, 'ig'), ''), replaceBy, key })
    }

    return prev
  }, source)
}
