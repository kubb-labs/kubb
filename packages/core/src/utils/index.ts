export type { FunctionParamsAST } from './FunctionParams.ts'
export { FunctionParams } from './FunctionParams.ts'
export {
  isPromise,
  isPromiseFulfilledResult,
  isPromiseRejectedResult,
} from './promise.ts'
export { renderTemplate } from './renderTemplate.ts'
export { timeout } from './timeout.ts'
export { getUniqueName, setUniqueName } from './uniqueName.ts'
export type { URLObject } from './URLPath.ts'
export { URLPath } from './URLPath.ts'
