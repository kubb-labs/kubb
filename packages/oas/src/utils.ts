import type { OASDocument, ParameterObject, SchemaObject } from 'oas/types'
import { isRef, isSchema } from 'oas/types'
import OASNormalize from 'oas-normalize'
import type { OpenAPI, OpenAPIV2, OpenAPIV3, OpenAPIV3_1 } from 'openapi-types'
import { isPlainObject, mergeDeep } from 'remeda'
import swagger2openapi from 'swagger2openapi'
import { Oas } from './Oas.ts'

export function isOpenApiV2Document(doc: any): doc is OpenAPIV2.Document {
  return doc && isPlainObject(doc) && !('openapi' in doc)
}
export function isOpenApiV3Document(doc: any): doc is OpenAPIV3.Document {
  return doc && isPlainObject(doc) && 'openapi' in doc
}

export function isOpenApiV3_1Document(doc: any): doc is OpenAPIV3_1.Document {
  return doc && isPlainObject<OpenAPIV3_1.Document>(doc) && 'openapi' in doc && doc.openapi.startsWith('3.1')
}

export function isJSONSchema(obj?: unknown): obj is SchemaObject {
  return !!obj && isSchema(obj)
}

export function isParameterObject(obj: ParameterObject | SchemaObject): obj is ParameterObject {
  return obj && 'in' in obj
}

/**
 * Determines if a schema is nullable, considering both the standard `nullable` property and the legacy `x-nullable` extension.
 *
 * @param schema - The schema object to check.
 * @returns `true` if the schema is marked as nullable; otherwise, `false`.
 */
export function isNullable(schema?: SchemaObject & { 'x-nullable'?: boolean }): boolean {
  return schema?.nullable ?? schema?.['x-nullable'] ?? false
}

/**
 * Determines if the given object is an OpenAPI ReferenceObject.
 *
 * @returns True if {@link obj} is a ReferenceObject; otherwise, false.
 */
export function isReference(obj?: any): obj is OpenAPIV3.ReferenceObject | OpenAPIV3_1.ReferenceObject {
  return !!obj && isRef(obj)
}

/**
 * Determines if the given object is a SchemaObject with a discriminator property of type DiscriminatorObject.
 *
 * @returns True if {@link obj} is a SchemaObject containing a non-string {@link discriminator} property.
 */
export function isDiscriminator(obj?: any): obj is SchemaObject & { discriminator: OpenAPIV3.DiscriminatorObject } {
  return !!obj && obj?.['discriminator'] && typeof obj.discriminator !== 'string'
}

/**
 * Determines whether a schema is required.
 *
 * Returns true if the schema has a non-empty {@link SchemaObject.required} array or a truthy {@link SchemaObject.required} property.
 *
 * @param schema - The schema object to check.
 * @returns True if the schema is required; otherwise, false.
 */
export function isRequired(schema?: SchemaObject): boolean {
  if (!schema) {
    return false
  }

  return Array.isArray(schema.required) ? !!schema.required?.length : !!schema.required
}

export function isOptional(schema?: SchemaObject): boolean {
  return !isRequired(schema)
}

export async function parse(
  pathOrApi: string | OASDocument,
  { oasClass = Oas, canBundle = true, enablePaths = true }: { oasClass?: typeof Oas; canBundle?: boolean; enablePaths?: boolean } = {},
): Promise<Oas> {
  const { loadConfig, bundle } = await import('@redocly/openapi-core')

  if (typeof pathOrApi === 'string' && canBundle) {
    // resolve external refs
    const config = await loadConfig()
    const bundleResults = await bundle({ ref: pathOrApi, config, base: pathOrApi })

    return parse(bundleResults.bundle.parsed as string)
  }

  const oasNormalize = new OASNormalize(pathOrApi, {
    enablePaths,
    colorizeErrors: true,
  })
  const document = (await oasNormalize.load()) as OpenAPI.Document

  if (isOpenApiV2Document(document)) {
    const { openapi } = await swagger2openapi.convertObj(document, {
      anchors: true,
    })

    return new oasClass({ oas: openapi as OASDocument })
  }

  return new oasClass({ oas: document })
}

export async function merge(pathOrApi: Array<string | OASDocument>, { oasClass = Oas }: { oasClass?: typeof Oas } = {}): Promise<Oas> {
  const instances = await Promise.all(pathOrApi.map((p) => parse(p, { oasClass, enablePaths: false, canBundle: false })))

  if (instances.length === 0) {
    throw new Error('No OAS instances provided for merging.')
  }

  const merged = instances.reduce(
    (acc, current) => {
      return mergeDeep(acc, current.document as OASDocument)
    },
    {
      openapi: '3.0.0',
      info: {
        title: 'Merged API',
        version: '1.0.0',
      },
      paths: {},
      components: {
        schemas: {},
      },
    } as any,
  )

  return parse(merged, { oasClass })
}
