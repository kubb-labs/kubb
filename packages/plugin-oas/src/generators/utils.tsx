import type { PluginFactoryOptions } from '@kubb/core'
import type { Operation, SchemaObject } from '@kubb/oas'
import { App, createReactFabric, type Fabric } from '@kubb/react-fabric'
import { Oas } from '../components/Oas.tsx'
import type { OperationGenerator } from '../OperationGenerator.ts'
import type { SchemaGenerator, SchemaGeneratorOptions } from '../SchemaGenerator.ts'
import type { Schema } from '../SchemaMapper.ts'
import type { ReactGenerator } from './/createReactGenerator.ts'

type BuildOperationsOptions<TOptions extends PluginFactoryOptions> = {
  fabric: Fabric
  generator: ReactGenerator<any>
  /**
   * @deprecated
   */
  instance: Omit<OperationGenerator<TOptions>, 'build'>
  options: TOptions['resolvedOptions']
}

export async function buildOperations<TOptions extends PluginFactoryOptions>(
  operations: Array<Operation>,
  { fabric, options, instance, generator }: BuildOperationsOptions<TOptions>,
): Promise<void> {
  if (!generator.Operations) {
    return undefined
  }

  const { pluginManager, oas, plugin, mode } = instance.context

  function Component() {
    return (
      <App meta={{ pluginManager, plugin, mode, oas }}>
        <Oas generator={instance}>
          <generator.Operations operations={operations} instance={instance} options={options} />
        </Oas>
      </App>
    )
  }

  const fabricChild = createReactFabric()
  fabricChild.render(Component)

  await fabric.context.fileManager.add(...fabricChild.files)
}

type BuildOperationOptions<TOptions extends PluginFactoryOptions> = BuildOperationsOptions<TOptions>

export async function buildOperation<TOptions extends PluginFactoryOptions>(
  operation: Operation,
  { fabric, options, instance, generator }: BuildOperationOptions<TOptions>,
): Promise<void> {
  if (!generator.Operation) {
    return undefined
  }

  const { pluginManager, oas, plugin, mode } = instance.context

  function Component() {
    return (
      <App meta={{ pluginManager, plugin: { ...plugin, options }, mode, oas }}>
        <Oas generator={instance}>
          <Oas.Operation operation={operation}>
            <generator.Operation operation={operation} options={options} instance={instance} />
          </Oas.Operation>
        </Oas>
      </App>
    )
  }
  const fabricChild = createReactFabric()
  fabricChild.render(Component)

  await fabric.context.fileManager.add(...fabricChild.files)
}

type BuildSchemaOptions<TOptions extends PluginFactoryOptions> = {
  fabric: Fabric
  generator: ReactGenerator<any>
  /**
   * @deprecated
   */
  instance: Omit<SchemaGenerator<SchemaGeneratorOptions, TOptions>, 'build'>
  options: TOptions['resolvedOptions']
}

export async function buildSchema<TOptions extends PluginFactoryOptions>(
  schema: {
    name: string
    tree: Array<Schema>
    value: SchemaObject
  },
  { fabric, options, instance, generator }: BuildSchemaOptions<TOptions>,
): Promise<void> {
  if (!generator.Schema) {
    return undefined
  }

  const { pluginManager, oas, plugin, mode } = instance.context

  function Component() {
    return (
      <App meta={{ pluginManager, plugin: { ...plugin, options }, mode, oas }}>
        <Oas>
          <Oas.Schema name={schema.name} schemaObject={schema.value} tree={schema.tree}>
            <generator.Schema schema={schema} options={options} instance={instance} />
          </Oas.Schema>
        </Oas>
      </App>
    )
  }
  const fabricChild = createReactFabric()
  fabricChild.render(Component)

  await fabric.context.fileManager.add(...fabricChild.files)
}
