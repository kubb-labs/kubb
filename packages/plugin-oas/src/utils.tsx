import type { Plugin, PluginFactoryOptions } from '@kubb/core'
import type { Operation, SchemaObject } from '@kubb/oas'
import { App, createReactFabric, type Fabric } from '@kubb/react-fabric'
import type { ReactGenerator } from './generators/createReactGenerator.ts'
import type { OperationGenerator } from './OperationGenerator.ts'
import type { SchemaGenerator, SchemaGeneratorOptions } from './SchemaGenerator.ts'
import type { Schema } from './SchemaMapper.ts'

type BuildOperationsOptions<TOptions extends PluginFactoryOptions> = {
  fabric: Fabric
  Component: ReactGenerator<any>['Operations']
  generator: Omit<OperationGenerator<TOptions>, 'build'>
  plugin: Plugin<TOptions>
}

export async function buildOperations<TOptions extends PluginFactoryOptions>(
  operations: Array<Operation>,
  { fabric, plugin, generator, Component }: BuildOperationsOptions<TOptions>,
): Promise<void> {
  if (!Component) {
    return undefined
  }

  const { pluginManager, oas, mode } = generator.context

  const fabricChild = createReactFabric()
  fabricChild.render(() => {
    return (
      <App meta={{ pluginManager, plugin, mode, oas }}>
        <Component operations={operations} generator={generator} plugin={plugin} />
      </App>
    )
  })

  await fabric.context.fileManager.add(...fabricChild.files)
}

type BuildOperationOptions<TOptions extends PluginFactoryOptions> = {
  fabric: Fabric
  Component: ReactGenerator<any>['Operation']
  generator: Omit<OperationGenerator<TOptions>, 'build'>
  plugin: Plugin<TOptions>
}

export async function buildOperation<TOptions extends PluginFactoryOptions>(
  operation: Operation,
  { fabric, plugin, generator, Component }: BuildOperationOptions<TOptions>,
): Promise<void> {
  if (!Component) {
    return undefined
  }

  const { pluginManager, oas, mode } = generator.context

  const fabricChild = createReactFabric()
  fabricChild.render(() => {
    return (
      <App meta={{ pluginManager, plugin, mode, oas }}>
        <Component operation={operation} plugin={plugin} generator={generator} />
      </App>
    )
  })

  await fabric.context.fileManager.add(...fabricChild.files)
}

type BuildSchemaOptions<TOptions extends PluginFactoryOptions> = {
  fabric: Fabric
  Component: ReactGenerator<any>['Schema']
  generator: Omit<SchemaGenerator<SchemaGeneratorOptions, TOptions>, 'build'>
  plugin: Plugin<TOptions>
}

export async function buildSchema<TOptions extends PluginFactoryOptions>(
  schema: {
    name: string
    tree: Array<Schema>
    value: SchemaObject
  },
  { fabric, plugin, Component, generator }: BuildSchemaOptions<TOptions>,
): Promise<void> {
  if (!Component) {
    return undefined
  }

  const { pluginManager, oas, mode } = generator.context

  const fabricChild = createReactFabric()
  fabricChild.render(() => {
    return (
      <App meta={{ pluginManager, plugin, mode, oas }}>
        <Component schema={schema} plugin={plugin} generator={generator} />
      </App>
    )
  })

  await fabric.context.fileManager.add(...fabricChild.files)
}
