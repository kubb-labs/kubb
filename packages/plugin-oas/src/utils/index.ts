export { getComments } from './getComments.ts'
export { getGroupedByTagFiles } from './getGroupedByTagFiles.ts'
export { getASTParams, getPathParams } from './getParams.ts'
export { getSchemaFactory } from './getSchemaFactory.ts'
export type { GetSchemasProps } from './getSchemas.ts'
export { getSchemas } from './getSchemas.ts'
export { refsSorter } from './refSorter.ts'
export { parseFromConfig } from './parseFromConfig.ts'
