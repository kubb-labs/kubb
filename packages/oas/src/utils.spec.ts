import { merge, parse } from './utils.ts'

describe('utils', () => {
  test('merge of 2 oas documents', async () => {
    const documents = [
      `openapi: 3.0.0
info:
  title: Swagger Petstore
  version: 1.0.0
components:
  schemas:
    Point:
      type: object
      properties:
        x:
          type: number
        y:
          type: number
      required: [x, y]
`,
      `openapi: 3.0.0
info:
  title: Shapes
  version: 1.0.0
paths: {}
components:
  schemas:
    Square:
      type: object
      properties:
        topLeft:
          $ref: '#/components/schemas/Point'
        size:
          type: number
      required: [topLeft, size]`,
    ]
    const oas = await merge(documents)

    expect(oas).toBeDefined()
    expect(oas.document).toMatchSnapshot()
    expect(oas.api?.info.title).toBe('Shapes')
  })

  test('parse a simple oas document', async () => {
    const oas = await parse(
      `openapi: 3.0.0
info:
  title: Swagger Petstore
  version: 1.0.0
components:
  schemas:
    Point:
      type: object
      properties:
        x:
          type: number
        y:
          type: number
      required: [x, y]
`,
      { canBundle: false },
    )
    expect(oas.api?.info.title).toBe('Swagger Petstore')
  })
})
