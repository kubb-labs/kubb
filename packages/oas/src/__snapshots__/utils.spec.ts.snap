// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > merge of 2 oas documents 1`] = `
{
  "components": {
    "schemas": {
      "Point": {
        "properties": {
          "x": {
            "type": "number",
          },
          "y": {
            "type": "number",
          },
        },
        "required": [
          "x",
          "y",
        ],
        "type": "object",
      },
      "Square": {
        "properties": {
          "size": {
            "type": "number",
          },
          "topLeft": {
            "$ref": "#/components/schemas/Point",
          },
        },
        "required": [
          "topLeft",
          "size",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "title": "Shapes",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {},
}
`;
